pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -312,429 +316,566 @@ def newpkgcb(pkgId, name, arch):
                     optimize_pb.done = 1
                     optimize_pb.save()
                     return
 
-            self.repository.last_sync_revision_number = repomd.revision
+            self.repository.last_sync_revision_number = self.data.repomd.revision
 
-            if self.treeinfo:
-                d_artifacts = [
-                    DeclarativeArtifact(
-                        artifact=Artifact(),
-                        url=urljoin(remote_url, self.treeinfo["filename"]),
-                        relative_path=".treeinfo",
-                        remote=self.remote,
-                        deferred_download=False,
-                    )
-                ]
-                for path, checksum in self.treeinfo["download"]["images"].items():
-                    artifact = Artifact(**checksum)
-                    da = DeclarativeArtifact(
-                        artifact=artifact,
-                        url=urljoin(remote_url, path),
-                        relative_path=path,
-                        remote=self.remote,
-                        deferred_download=self.deferred_download
-                    )
-                    d_artifacts.append(da)
+            await self.parse_distribution_tree()
+
+            await self.parse_repository_metadata()
+            await self.parse_modules_metadata()
+            await self.parse_packages_components()
+            await self.parse_content()
+
+            # now send modules down the pipeline since all relations have been set up
+            for modulemd in self.data.modulemd_list:
+                await self.put(modulemd)
 
-                distribution_tree = DistributionTree(**self.treeinfo["distribution_tree"])
-                dc = DeclarativeContent(content=distribution_tree, d_artifacts=d_artifacts)
-                dc.extra_data = self.treeinfo
+            for dc_group in self.data.dc_groups:
+                await self.put(dc_group)
+
+    def should_optimize_sync(self):
+        """Check whether it is possible to optimize the synchronization or not."""
+        # Caution: we are not storing when the remote was last updated, so the order of this
+        # logic must remain in this order where we first check the version number as other
+        # changes than sync could have taken place such that the date or repo version will be
+        # different from last sync
+        return (self.optimize and self.repository.last_sync_remote and
+                self.remote.pk == self.repository.last_sync_remote.pk and
+                (self.repository.last_sync_repo_version ==
+                 self.repository.latest_version().number) and
+                (self.remote.pulp_last_updated <=
+                 self.repository.latest_version().pulp_created) and
+                is_previous_version(self.data.repomd.revision,
+                                    self.repository.last_sync_revision_number)
+                )
```

Comment:
I would consider using nested if statements to break this logic up rather than having it all in one massive conditional.  You don't have to do this, but try it just to see if it ends up looking cleaner.  If it isn't, just keep it this way.

Reply:
I just fixed the indentation. I am afraid to break this logic up due to the warning above.